### PR TITLE
Batch sync, default and continuous lanes

### PR DESCRIPTION
**Pull request by tyao1**
## Summary

This is the other approach for unifying default and sync lane https://github.com/facebook/react/pull/25524.
The approach in that PR is to merge default and continuous lane into the sync lane, and use a new field to track the priority. But there are a couple places that field will be needed, and it is difficult to correctly reset the field when there is no sync lane.

In this PR we take the other approach that doesn't remove any lane, but batch them to get the behavior we want.

## How did you test this change?

yarn test

### CODE REVIEW

**Comment by Andarist** [comment]
Is this going to fix those two?
https://github.com/facebook/react/issues/24831
https://github.com/facebook/react/issues/25191

File: packages/react-reconciler/src/ReactFiberLane.new.js [hunk 1]
```diff
@@ -135,8 +136,28 @@ let nextRetryLane: Lane = RetryLane1;
 function getHighestPriorityLanes(lanes: Lanes | Lane): Lanes {
   switch (getHighestPriorityLane(lanes)) {
     case SyncHydrationLane:
+      if (enableUnifiedSyncLane) {
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberLane.new.js:139]: To simplify this logic a bit, let's just put a single branch at the top of this function:

```js
if (enableUnifiedSyncLane) {
  // This is a constant that includes Default, Sync, etc. But not hydration lanes.
  const pendingSyncLanes = lanes & SyncUpdateLanes;
  if (pendingSyncLanes !== 0) {
    return pendingSyncLanes;
  }
}
```

We can handle the hydration lanes in a similar way in `getBumpedLaneForHydration`:

```js
export function getBumpedLaneForHydration(
  root: FiberRoot,
  renderLanes: Lanes,
): Lane {
  if (enableUnifiedSyncLane) {
    const pendingSyncLanes = lanes & SyncUpdateLanes;
    if (pendingSyncLanes !== 0) {
      return SyncHydrationLane;
    }
  }
}
 
```

**Comment by acdlite** [comment]
@Andarist That's one of the motivations, yeah.

File: packages/react-reconciler/src/ReactFiberLane.js [hunk 2]
```diff
@@ -251,7 +260,10 @@ export function getNextLanes(root: FiberRoot, wipLanes: Lanes): Lanes {
       // Default priority updates should not interrupt transition updates. The
       // only difference between default updates and transition updates is that
       // default updates do not support refresh transitions.
-      (nextLane === DefaultLane && (wipLane & TransitionLanes) !== NoLanes)
+      // Interrupt transtion if default is batched with sync.
+      (!enableUnifiedSyncLane &&
+        nextLane === DefaultLane &&
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberLane.js:265]: Can you remind me why you changed this?

The desired behavior is that DefaultLane should only interrupt a transition if there's also a pending SyncLane that's entangled with. But if they are entangled, then the `nextLane >= wipLane` check above would have already evaluated to true.

So I don't think this condition needs to change.

Either way, we should probably have a test for this.

File: packages/react-reconciler/src/ReactFiberLane.js [hunk 3]
```diff
@@ -754,46 +766,53 @@ export function getBumpedLaneForHydration(
...
-      lane = IdleHydrationLane;
-      break;
-    default:
-      // Everything else is already either a hydration lane, or shouldn't
-      // be retried at a hydration lane.
-      lane = NoLane;
-      break;
+    }
+  }
+  if (!lane) {
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberLane.js:774]: Use an `else` instead:

```js
let lane;
if (enableUnifiedSyncLane && (renderLane & SyncUpdateLanes) !== NoLane) {
  lane = SyncHydrationLane;
} else  {
  ...
}

**Review by acdlite — APPROVED** [review]
Overall looks great! Just a few more questions. Could you please look at them before merging?

**Comment by Siegrift** [comment]
Based on https://github.com/facebook/react/pull/25700#issuecomment-1330963778

Should these two issues be closed?
https://github.com/facebook/react/issues/24831
https://github.com/facebook/react/issues/25191

**Pull request merged**